pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -320,34 +324,61 @@ func (l *LibvirtDomainManager) FinalizeVirtualMachineMigration(vmi *v1.VirtualMa
 	return l.finalizeMigrationTarget(vmi)
 }
 
-// HotplugHostDevices attach host-devices to running domain
-// Currently only SRIOV host-devices are supported
+// HotplugHostDevices attach host-devices to running domain, currently only SRIOV host-devices are supported.
+// This is operation runs on the background each at a time.
 func (l *LibvirtDomainManager) HotplugHostDevices(vmi *v1.VirtualMachineInstance) error {
+	if l.hotplugHostDevicesInProgress.Load().(bool) {
+		log.Log.V(3).Object(vmi).Info("hot-plug host-devices is in progress")
+		return nil
+	}
+
+	l.asyncHotplugHostDevices(vmi)
+
+	return nil
+}
+
+func (l *LibvirtDomainManager) asyncHotplugHostDevices(vmi *v1.VirtualMachineInstance) {
+	l.hotplugHostDevicesInProgress.Store(true)
+
+	go func() {
+		defer l.hotplugHostDevicesInProgress.Store(false)
+
+		log.Log.V(3).Object(vmi).Info("starting hot-plug host-devices")
+
+		if err := l.hotplugHostDevices(vmi); err != nil {
+			log.Log.Object(vmi).Error(err.Error())
+		}
+	}()
+}
+
+func (l *LibvirtDomainManager) hotplugHostDevices(vmi *v1.VirtualMachineInstance) error {
 	l.domainModifyLock.Lock()
 	defer l.domainModifyLock.Unlock()
 
+	const errMsgPrefix = "failed to hot-plug host-devices"
+
 	domainName := api.VMINamespaceKeyFunc(vmi)
 	domain, err := l.virConn.LookupDomainByName(domainName)
 	if err != nil {
-		return err
+		return fmt.Errorf("%s: %v", errMsgPrefix, err)
 	}
 	defer domain.Free()
 
 	domainSpec, err := util.GetDomainSpecWithFlags(domain, 0)
 	if err != nil {
-		return err
+		return fmt.Errorf("%s: %v", errMsgPrefix, err)
 	}
 
 	sriovHostDevices, err := sriov.GetHostDevicesToAttach(vmi, domainSpec)
 	if err != nil {
-		return err
+		return fmt.Errorf("%s: %v", errMsgPrefix, err)
 	}
 
 	if err := hostdevice.AttachHostDevices(domain, sriovHostDevices); err != nil {
-		return err
+		return fmt.Errorf("%s: %v", errMsgPrefix, hostdevice.AttachHostDevices(domain, sriovHostDevices))
 	}
 
-	return nil
+	return err
```

Comment:
Why do you return err here and not nil?

Reply:
Thanks for the heads up, I changed it to nil.